File: SRM 638/ShadowSculpture.cpp
#include <cstdio>
#include <cmath>
#include <cstring>
#include <ctime>
#include <iostream>
#include <algorithm>
#include <set>
#include <queue>
#include <vector>
#include <sstream>
#include <typeinfo>
#include <fstream>

using namespace std;
const int M = 15;
int a[M][M][M];
int g[M][M][M];
#define REP(i,a) for(int i=0;i<(a);++i)
#define IN(i,a) (0 <= i && i < a)
class ShadowSculpture {
    int n;
    vector<string> XY;
    vector<string> YZ;
    vector<string> ZX;
    bool check (int x, int y, int z) {
        if (!(0 <= x && x < n)) return false;
        if (!(0 <= y && y < n)) return false;
        if (!(0 <= z && z < n)) return false;
        if (XY[x][y] == 'N') return false;
        if (YZ[y][z] == 'N') return false;
        if (ZX[z][x] == 'N') return false;
        return true;
    }
    public:
    string possible(vector<string> _XY, vector<string> _YZ, vector<string> _ZX) {
        XY = _XY;
        YZ = _YZ;
        ZX = _ZX;
        int dx[6] = {1, -1, 0, 0, 0, 0};
        int dy[6] = {0, 0, 1, -1, 0, 0};
        int dz[6] = {0, 0, 0, 0, 1, -1};
        n = XY.size();
        REP(sx, n) REP(sy, n) REP(sz, n) if(check(sx, sy, sz)) {
            //cout << sx << " " << sy << " " << sz << endl;
            queue<int> qx, qy, qz;
            qx.push(sx);
            qy.push(sy);
            qz.push(sz);
            bool used[10][10][10] = {};
            used[sx][sy][sz] = true;
            while(!qx.empty()) {
                int x = qx.front();
                int y = qy.front();
                int z = qz.front();
                qx.pop();
                qy.pop();
                qz.pop();
                REP(r, 6) {
                    int nx = x + dx[r];
                    int ny = y + dy[r];
                    int nz = z + dz[r];
                    if(check(nx, ny, nz) && !used[nx][ny][nz]) {
                        used[nx][ny][nz] = true;
                        qx.push(nx);
                        qy.push(ny);
                        qz.push(nz);
                    }
                }
            }
            //REP(x, n) REP(y, n) REP(z, n) printf("(%d, %d, %d) %d\n", x, y, z, used[x][y][z]);
 
            bool XYB[10][10] = {};
            bool YZB[10][10] = {};
            bool ZXB[10][10] = {};
            REP(x, n) REP(y, n) REP(z, n) if(used[x][y][z]) {
                XYB[x][y] = true;
                YZB[y][z] = true;
                ZXB[z][x] = true;
            }
 
            bool ok = true;
            REP(x, n)REP(y, n) if(XY[x][y] == 'Y' && !XYB[x][y]) ok = false;
            REP(y, n)REP(z, n) if(YZ[y][z] == 'Y' && !YZB[y][z]) ok = false;
            REP(z, n)REP(x, n) if(ZX[z][x] == 'Y' && !ZXB[z][x]) ok = false;
            if(ok) return "Possible";
        }
 
        bool XYB[10][10] = {};
        bool YZB[10][10] = {};
        bool ZXB[10][10] = {};
        bool ok = true;
        REP(x, n)REP(y, n) if(XY[x][y] == 'Y' && !XYB[x][y]) ok = false;
        REP(y, n)REP(z, n) if(YZ[y][z] == 'Y' && !YZB[y][z]) ok = false;
        REP(z, n)REP(x, n) if(ZX[z][x] == 'Y' && !ZXB[z][x]) ok = false;
        if(ok) return "Possible";
        return "Impossible";
    }
};

// CUT begin
ifstream data("ShadowSculpture.sample");

string next_line() {
    string s;
    getline(data, s);
    return s;
}

template <typename T> void from_stream(T &t) {
    stringstream ss(next_line());
    ss >> t;
}

void from_stream(string &s) {
    s = next_line();
}

template <typename T> void from_stream(vector<T> &ts) {
    int len;
    from_stream(len);
    ts.clear();
    for (int i = 0; i < len; ++i) {
        T t;
        from_stream(t);
        ts.push_back(t);
    }
}

template <typename T>
string to_string(T t) {
    stringstream s;
    s << t;
    return s.str();
}

string to_string(string t) {
    return "\"" + t + "\"";
}

bool do_test(vector<string> XY, vector<string> YZ, vector<string> ZX, string __expected) {
    time_t startClock = clock();
    ShadowSculpture *instance = new ShadowSculpture();
    string __result = instance->possible(XY, YZ, ZX);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << to_string(__expected) << endl;
        cout << "           Received: " << to_string(__result) << endl;
        return false;
    }
}

int run_test(bool mainProcess, const set<int> &case_set, const string command) {
    int cases = 0, passed = 0;
    while (true) {
        if (next_line().find("--") != 0)
            break;
        vector<string> XY;
        from_stream(XY);
        vector<string> YZ;
        from_stream(YZ);
        vector<string> ZX;
        from_stream(ZX);
        next_line();
        string __answer;
        from_stream(__answer);

        cases++;
        if (case_set.size() > 0 && case_set.find(cases - 1) == case_set.end())
            continue;

        cout << "  Testcase #" << cases - 1 << " ... ";
        if ( do_test(XY, YZ, ZX, __answer)) {
            passed++;
        }
    }
    if (mainProcess) {
        cout << endl << "Passed : " << passed << "/" << cases << " cases" << endl;
        int T = time(NULL) - 1502070009;
        double PT = T / 60.0, TT = 75.0;
        cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
        cout << "Score  : " << 300 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
    set<int> cases;
    bool mainProcess = true;
    for (int i = 1; i < argc; ++i) {
        if ( string(argv[i]) == "-") {
            mainProcess = false;
        } else {
            cases.insert(atoi(argv[i]));
        }
    }
    if (mainProcess) {
        cout << "ShadowSculpture (300 Points)" << endl << endl;
    }
    return run_test(mainProcess, cases, argv[0]);
}
// CUT end

File: SRM 638/ShadowSculpture.java
import java.io.*;
import java.util.*;

public class ShadowSculpture {
    private int n;
    private String[] XY;
    private String[] YZ;
    private String[] ZX;
    private int[] dx = {1, 0, 0, 0, 0, -1}; 
    private int[] dy = {0, 1, 0, -1, 0, 0}; 
    private int[] dz = {0, 0, -1, 0, 1, 0}; 
    boolean check(int x, int y, int z) {
        if (x < 0 || x >= n) return false;
        if (y < 0 || y >= n) return false;
        if (z < 0 || z >= n) return false;
        if (XY[x].charAt(y) == 'N') return false;
        if (YZ[y].charAt(z) == 'N') return false;
        if (ZX[z].charAt(x) == 'N') return false;
        return true;

    }
	public String possible(String[] XY, String[] YZ, String[] ZX) {
        n = XY.length;
        this.XY = XY;
        this.YZ = YZ;
        this.ZX = ZX;
        boolean flag = false;
        for (int i = 0; i < n; ++i) for (int j = 0; j < n; ++j) {
            if (XY[i].charAt(j) == 'Y') flag = true;
            if (YZ[i].charAt(j) == 'Y') flag = true;
            if (ZX[i].charAt(j) == 'Y') flag = true;
        }
        if (!flag) return "Possible";
        for (int x = 0; x < n; ++x) for (int y = 0; y < n; ++y) for (int z = 0; z < n; ++z) if (check(x, y, z)) {
            ArrayDeque<Integer>Q = new ArrayDeque<Integer>();
            Q.add(x);
            Q.add(y);
            Q.add(z);
            boolean[][][] used = new boolean[n][n][n];
            used[x][y][z] = true;
            while (Q.size() > 0) {
                int cx = Q.remove();
                int cy = Q.remove();
                int cz = Q.remove();
                for (int dir = 0; dir < 6; ++dir) {
                    int nx = cx + dx[dir];
                    int ny = cy + dy[dir];
                    int nz = cz + dz[dir];
                    if (check(nx, ny, nz) && !used[nx][ny][nz]) {
                        used[nx][ny][nz] = true;
                        Q.add(nx);
                        Q.add(ny);
                        Q.add(nz);
                    }
                }
            }
            boolean possible = true;
            for (int i = 0; i < n; ++i) for (int j = 0; j < n; ++j) {
                if (XY[i].charAt(j) == 'Y') {
                    boolean ok = false;
                    for (int k = 0; k < n; ++k) if (used[i][j][k]) ok = true;
                    if (!ok) possible = false;
                }
                if (YZ[i].charAt(j) == 'Y') {
                    boolean ok = false;
                    for (int k = 0; k < n; ++k) if (used[k][i][j]) ok = true;
                    if (!ok) possible = false;
                }
                if (ZX[i].charAt(j) == 'Y') {
                    boolean ok = false;
                    for (int k = 0; k < n; ++k) if (used[j][k][i]) ok = true;
                    if (!ok) possible = false;
                }
            }
            if (possible) return "Possible";
        }
		return "Impossible";
	}

// CUT begin
	public static void main(String[] args){
		System.err.println("ShadowSculpture (300 Points)");
		System.err.println();
		HashSet<Integer> cases = new HashSet<Integer>();
        for (int i = 0; i < args.length; ++i) cases.add(Integer.parseInt(args[i]));
        runTest(cases);
	}

	static void runTest(HashSet<Integer> caseSet) {
	    int cases = 0, passed = 0;
	    while (true) {
	    	String label = Reader.nextLine();
	    	if (label == null || !label.startsWith("--"))
	    		break;

            String[] XY = new String[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < XY.length; ++i)
                XY[i] = Reader.nextLine();
            String[] YZ = new String[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < YZ.length; ++i)
                YZ[i] = Reader.nextLine();
            String[] ZX = new String[Integer.parseInt(Reader.nextLine())];
            for (int i = 0; i < ZX.length; ++i)
                ZX[i] = Reader.nextLine();
            Reader.nextLine();
            String __answer = Reader.nextLine();

            cases++;
            if (caseSet.size() > 0 && !caseSet.contains(cases - 1))
                continue;
    		System.err.print(String.format("  Testcase #%d ... ", cases - 1));

            if (doTest(XY, YZ, ZX, __answer))
                passed++;
	    }
	    if (caseSet.size() > 0) cases = caseSet.size();
        System.err.println(String.format("%nPassed : %d/%d cases", passed, cases));

        int T = (int)(System.currentTimeMillis() / 1000) - 1502066821;
        double PT = T / 60.0, TT = 75.0;
        System.err.println(String.format("Time   : %d minutes %d secs%nScore  : %.2f points", T / 60, T % 60, 300 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))));
	}

	static boolean doTest(String[] XY, String[] YZ, String[] ZX, String __expected) {
		for (int i = 0; i < XY.length; i++) {
			XY[i] = new String(XY[i]);
		}
		for (int i = 0; i < YZ.length; i++) {
			YZ[i] = new String(YZ[i]);
		}
		for (int i = 0; i < ZX.length; i++) {
			ZX[i] = new String(ZX[i]);
		}
		long startTime = System.currentTimeMillis();
		Throwable exception = null;
		ShadowSculpture instance = new ShadowSculpture();
		String __result = "";
		try {
			__result = instance.possible(XY, YZ, ZX);
		}
		catch (Throwable e) { exception = e; }
		double elapsed = (System.currentTimeMillis() - startTime) / 1000.0;

		if (exception != null) {
			System.err.println("RUNTIME ERROR!");
			exception.printStackTrace();
			return false;
		}
		else if (__expected.equals(__result)) {
			System.err.println("PASSED! " + String.format("(%.2f seconds)", elapsed));
			return true;
		}
		else {
			System.err.println("FAILED! " + String.format("(%.2f seconds)", elapsed));
			System.err.println("           Expected: " + "\"" + __expected + "\"");
			System.err.println("           Received: " + "\"" + __result + "\"");
			return false;
		}
	}

	static class Reader {
        private static final String dataFileName = "ShadowSculpture.sample";
	    private static BufferedReader reader;

	    public static String nextLine() {
	        try {
                if (reader == null) {
                    reader = new BufferedReader(new InputStreamReader(new FileInputStream(dataFileName)));
                }
                return reader.readLine();
	        }
	        catch (IOException e) {
	            System.err.println("FATAL!! IOException");
	            e.printStackTrace();
	            System.exit(1);
	        }
	        return "";
	    }
	}
// CUT end
}
